provider/image.go

```diff
@@ -264,6 +265,35 @@ func (p *dockerNativeProvider) dockerBuild(ctx context.Context,
 		opts.SessionID = sess.ID()
 	}
 
+	// Docker with BuildKit seems to not cache multi-stage builds every other build. This is a known issue:
+	// - https://github.com/moby/buildkit/issues/3730
+	// - https://github.com/moby/buildkit/issues/2274
+	// - https://github.com/moby/buildkit/issues/1981
+	//
+	// A workaround suggested in these threads is to `docker pull` the image prior to building.
+
+	// In this loop, we pull the cached images, and no errors inside this loop are fatal. If we fail
+	// to pull an image, we log a warning and continue.
+	for _, cachedImage := range img.Build.CachedImages {
+		auth, msg, err := getRegistryAuth(img, cfg)
+		if err != nil {
+			err = p.host.Log(ctx, "warning", urn, msg)
+			continue
+		}
+		if msg != "" {
+			err = p.host.Log(ctx, "warning", urn, msg)
+			if err != nil {
+				continue
+			}
+		}
```

Comment:
I think we don't need this block as past! us were smart and we should be [already logged in if cacheImages is set](https://github.com/pulumi/pulumi-docker/blob/master/provider/image.go#L219) - we can use regAuth for the call to `pullDockerImage`.

Reply:
Ah, I think as it's written it'll help us implement #497 when we get to it, no harm in keeping the code as it is?